Addressing issue #266 by providing a mapping of the data model for JWT

awoie opened the pull request:
This is an initial PR to address issue #266 by providing a mapping of the data model for JWT.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/awoie/vc-data-model/pull/267.html" title="Last updated on Dec 15, 2018, 8:24 PM UTC (7358723)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/267/a7ec4d4...awoie:7358723.html" title="Last updated on Dec 15, 2018, 8:24 PM UTC (7358723)">Diff</a>

awoie commented:
I'm a member of the W3C VC WG. Could someone please fix this?

Thanks,
Oliver

awoie commented:
I'm a member of W3C VC WG and I've connected my W3C account to my Github handle. Could some please provide some help on how to fix this?

Thanks,
Oliver

msporny commented on index.html:
```diff
@@ -2187,7 +2191,7 @@ <h2>Syntaxes</h2>
 (described in Sections <a href="#core-data-model"></a>,
 <a href="#basic-concepts"></a>, and <a href="#advanced-concepts"></a>
-) is realized in JSON and JSON-LD. Although syntactic mappings are only provided for these two syntaxes,
+) is realized in JSON, JSON-LD and JWT. Although syntactic mappings are only provided for these three syntaxes,
```
Editorial: "JSON-LD, and" -- we do oxford comma 'round these parts. :)

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+            <ul>
+                <li>The signature of the JWT is produced by the issuer of the verifiable credential.</li>
+                <li>Because JWS implicitly provides a verifiable proof of the contained verifiable credentials, the "proof" is no longer needed. Please note that
+                    the "jti" claim represents a nonce value.</li>
```
There are two ways that we could go with this: Add the JWT keywords (jti, iat, etc.) to the JSON-LD Credentials vocab... or leave them out, and move the JSON-LD Context down one level.

The question this raises is: Is the top level just an encapsulating container? Or is it a VC? If it's  just a container, where does evidence and termsOfUse go? I expect we'll end up with the top level being a VC, which then begs the question of whether or not JSON-LD processing is required or not during verification to ensure the semantic mappings between the sending and receiving application are checked.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                <li>The signature of the JWT is produced by the issuer of the verifiable credential.</li>
+                <li>Because JWS implicitly provides a verifiable proof of the contained verifiable credentials, the "proof" is no longer needed. Please note that
+                    the "jti" claim represents a nonce value.</li>
+                <li>The IANA JWT claims registry should be enforced whenever possible:</li>
```
Does this mean that the VC Core Data Model properties should be added to the IANA JWT claims registry? What happens with `evidence` and `termsOfUse`, for example?

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                    <li>The "id" contained in the "claim" object becomes the JWT "sub".</li>
+                    <li>If "created" and "creator" are required, these claims could be added to the JWT payload.</li>
+                    <li>The "issuer" becomes the JWT "iss" claim.</li>
+                    <li>The "issuanceDate" becomes the JWT "iat" claim.</li>
```
Where is the nonce and domain specified? If you don't do that, the message can be replayed. This is confusing layers of abstraction... the VC Data Model spec isn't supposed to talk about the crypto... I don't think the group is chartered to do that. We may need to find non-normative weasel wording to make these sorts of assertions since certain W3C Member companies requested that we NOT make statements like the above.

awoie replied:
True, `nonce` has also to be included. Domain would be the `iss` claim. I am now aware of the group charter. It would be good to have a non-normative example similar to the JSON LD representation where the `nonce` is part of the `proof` attribute.

awoie replied:
I added the `nonce`to the VP and VC examples. Please note, the `jti`serves as a nonce in the VC already. In general, we don't talk about specific risk mitigation strategies and everybody is free to implement their own means using what the data model allows.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+  "jti": "urn:uuid:3978344f-8596-4c3a-a978-8fcaba3903c5",
+  "iss": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+  "aud": "https://dmv.example.gov/issuers/14",
+  "nonce": "343§$FSFDa-",
```
What prevents replay across multiple domains?

dlongley replied:
I think `aud` could be used to do that, but here it appears to be an issuer URL. But perhaps this VP is actually being sent to the same issuer that the VC was issued from? If so, it may confuse the reader.

awoie replied:
Yes, it is not the best example. The intention was to use `aud` for that. I will update the example.

awoie replied:
I updated the examples.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+}
+          </pre>
+
+            <p class="issue">
```
I'm supportive of pulling this PR in as long as:

1. There are multiple implementers - who are the multiple implementers for this PR?
2. We are very clear about the limitations of VCs encoded as JWTs - signatures are tied to a very specific serialization where whitespace destroys the signature, signatures are tied to JSON, base64 encoding makes developers have to use twice as much storage in the simples case, set and chained signatures are problematic, no support for other types of mathematical proof formats, if JSON-LD processing is not done there is no guarantee that the semantics of the message is well formed, etc. We need a very large warning section here so that developers know what they're getting themselves into when using JWTs.

Digital Bazaar's position is that combining VCs with JWTs are a footgun, we won't be implementing it, but we also don't want to stand in the way of other companies that still want to go down that path in spite of all of the warnings.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+
+            <pre class="example nohighlight" title="Verifiable credential JWT payload">
+{
+  "@context": [
```
What is to prevent the @context from using different values such that JSON-LD processors interpret the message differently?

awoie replied:
This is true. We might end up removing the `@context` from the JWT data model. Let's discuss that during our call.

awoie replied:
I added the changes we discussed during the our JWT call. I hope, this will resolve the discussion.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+  ],
+  "sub": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+  "jti": "http://dmv.example.gov/credentials/3732",
+  "iss": "https://dmv.example.gov/issuers/14",
```
How is the issuer key bound to the message?

msporny replied:
Yes, that should work, I think.

awoie replied:
In my mind, key discovery is out of scope of this specification. We are talking about a data model and not about protocols. People could implement that either by DID and DID Document resolution based on the `iss` or even based JWK. These would be part of a different spec.

msporny commented:
>> I'm supportive of pulling this PR in as long as:
>> 1. There are multiple implementers - who are the multiple implementers for this PR?
> 
> @msporny , that has not been a required bar to date for any feature, and this came in before feature freeze. That said, if there are insufficient implementations when we reach our final CR review date we will of course need to pull it out at that time, as with any of our features that are insufficiently implemented.

Yep, sorry, that's what I meant.

So, I suggest we:

1. Mark this as a feature at risk so we can get it into the spec.
2. Outline when it's safe to use the feature and when you may want to choose a different approach.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+  "iat": "1541493724",
+  "exp": "1573029723",
+  "type": ["VerifiablePresentation"],
+  "verifiableCredential": ["
```
This should probably change to `verifiableCredentialJwt` as the value is encoded.

awoie replied:
I updated the the JWT data model to reflect the outcome of our JWT discussion. The changes should address this issue.

msporny commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+    "https://example.com/examples/v1"
+  ],
+  "sub": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+  "jti": "http://dmv.example.gov/credentials/3732",
```
Just looked at rfc7519, `sub` and `jti` are optional, so they don't in and of themselves ensure correlation. There are no ZKP-style signature mechanisms for JWTs, so I'd be interested in hearing how they support the various types of non-correlation identified in the security/privacy section.

awoie replied:
One approach will be to add the `proof` to the JWT if something else than digital signatures is required. In this case, a ZKP representation could have no JWS but include the `proof` attribute.

awoie replied:
The PR does not support native JWT integration of ZKP, or complex proofs but however, we can include a JSON/JSON-LD verifiable credential/presentation, we would be able to make use of the `proof`.

msporny commented:
Given all of the discussion, I've tried to capture the sets of features JWTs and LD-Proofs enable: https://docs.google.com/spreadsheets/d/1WJ1pOH6zmZacuWgoAVfIHLI9Q1EPXbScY44w2gqv9vo/edit

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                <ul>
+                    <li>The "id" of the presentation becomes the JWT "jti" claim.</li>
+                    <li>A "nonce" claim MUST be included to mitigate the risk of replay attacks.</li>
+                    <li>A "aud" claim may be included and contains the subject of the verifier. The "aud" claim MUST be one or more URIs.</li>
```
My understanding was to not make any statements on security. Same applies to nonce. I will update the PR accordingly.

dlongley commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+            </p>
+
+            <p class="issue">
+                Add an example of how a base64 JWT encoded verifiable credential looks like.
```
`how` => `what`

dlongley commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                  "https://example.com/examples/v1"
+                ],
+                "type": ["VerifiablePresentation"],
+                "vc_jwt": ["..."]
```
`vc_jwt` should be `verifiableCredential`.

awoie replied:
agree

dlongley commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+              "iat": "1541493724",
+              "exp": "1573029723",
+              "nonce": "343§$FSFDa-",
+              "vcp": {
```
Should `vcp` be `vp` to match `vc`?

awoie replied:
agree

pelle commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                For backwards compatibility with JWT processors the following JWT registered claim names MUST be used
+                instead or in addition of their respective verifiable credential counterparts:
+                <ul>
+                    <li><code>exp</code> SHOULD represent expirationDate.</li>
```
I agree with MUST, but if applicable

awoie replied:
agree

pelle commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                to the use of JWTs for Verifiable Credentials. These concerns
+                will be documented in time in at least the Verfiable Claims Model
+                and Security Considerations section of this document.
+            </p>
```
@msporny I recommend having that in a separate document an not in the spec.

pelle commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                In addition, the following JOSE headers MUST be present in the JWS:
+                <ul>
+                    <li><code>alg</code> MUST be used for RSA and ECDSA based digital signatures.</li>
+                    <li><code>kid</code> MUST be used if there are multiple keys associated with the issuer of the JWT.
```
I don't think this is needed. It's an option but adds extra verbosity. Our implementation checks the DID document and tests each public key. This works fine in almost all use cases and removes duplication.

awoie replied:
- Agree with MUST. 
- I also would prefer to see this in a separate document. 
- `kid` MAY be used.

pelle replied:
👍

msporny commented:
@awoie, can you please respond to @pelle so we can get this merged?

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                  "https://example.com/examples/v1"
+                ],
+                "type": ["VerifiableCredential", "ProofOfAgeCredential"],
+                "credential_subject": {
```
I know that camelCase is the preferred way of naming attributes with JSON-LD. But I was wondering if it makes sense to define a context for snake_case, or even have shorter attribute names? During W3C, guys from the JWT community were a bit concerned about the size of the credential.

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+              "sub": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+              "jti": "http://dmv.example.gov/credentials/3732",
+              "iss": "did:example:abfe13f712120431c276e12ecab",
+              "iat": "1541493724",
```
I agree. Applies to `iat` and `exp`.

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                    <li><code>kid</code> MUST be used if there are multiple keys associated with the issuer of the JWT.
+                        Key discovery is out of scope of this specification. For example, the <code>kid</code> can
+                        refer to a key in a DID Document, or can be the identifier of a key inside a JWKS.</li>
+                    <li><code>typ</code> MUST be set to JWT.</li>
```
agree

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+        <section>
+            <h2>JSON Web Token</h2>
+            <p>
+                JSON Web Token (JWT) [[!RFC7519]] is still a widely used means to express claims to be transferred
```
agree

awoie commented on index.html:
```diff
@@ -2322,6 +2258,169 @@ <h2>JSON-LD</h2>
+                ],
+                "type": ["VerifiableCredential", "ProofOfAgeCredential"],
+                "credential_subject": {
+                  "age_over": 21
```
See my answer above.

awoie commented:
Answered the questions and updated the PR. Also addressed merge conflicts.

Thanks,
Oliver

> On 12. Dec 2018, at 14:34, Manu Sporny <notifications@github.com> wrote:
> 
> @awoie <https://github.com/awoie>, can you please respond to @pelle <https://github.com/pelle> so we can get this merged?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/w3c/vc-data-model/pull/267#issuecomment-446715222>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ApbdsCb7-kgG-qSI-xvhQyopcf1k99mXks5u4VozgaJpZM4YQLzU>.
>

davidlehn commented on index.html:
```diff
@@ -2787,9 +2787,10 @@ <h3>JWT and JWS Considerations</h3>
-    "credentialSubject": {
-      "ageOver": 21
+    "type": ["VerifiableCredential", "UniversityDegreeCredential"],
+    "degree": {
```
This VC should have a "credentialSubject" property that contains the "degree" property and an "id" that matches "sub" above.
